File: Top200/offer32-LevelOrder.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type Queue struct {
	arr  []*TreeNode
	size int
}

func (q *Queue) Push(t *TreeNode) {
	q.arr = append(q.arr, t)
	q.size++
}

func (q *Queue) Pop() *TreeNode {
	if q.size <= 0 {
		return nil
	}

	res := q.arr[0]
	q.arr = q.arr[1:q.size]
	q.size--

	return res
}

func (q *Queue) IsEmpty() bool {
	if q.size <= 0 {
		return true
	}

	return false
}

func NewQueue() *Queue {
	return &Queue{
		arr:  []*TreeNode{},
		size: 0,
	}
}

func levelOrder(root *TreeNode) []int {
	queue := NewQueue()
	res := []int{}

	if root == nil {
		return res
	}

	queue.Push(root)

	for !queue.IsEmpty() {
		node := queue.Pop()

		if node.Left != nil {
			queue.Push(node.Left)
		}

		if node.Right != nil {
			queue.Push(node.Right)
		}

		res = append(res, node.Val)
	}

	return res
}

File: Top200/offer32-LevelOrder.java
package Top200;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

class LevelOrder {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    public int[] levelOrder(TreeNode root) {
        if (root == null) {
            return new int[]{};
        }
    
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        List<Integer> res = new ArrayList<Integer>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            TreeNode temp = queue.poll();
            res.add(temp.val);

            if (temp.left != null) {
                queue.offer(temp.left);
            }

            if (temp.right != null) {
                queue.offer(temp.right);
            }
        }
        
        int size = res.size();
        int[] resArray = new int[size];
        for (int i = 0; i < size; i++) {
            resArray[i] = res.get(i);
        }

        return resArray;
    }
    
}
